pkg/labels/fake_applicator.go

```diff
@@ -102,21 +104,124 @@ func (app *fakeApplicator) GetMatches(selector labels.Selector, labelType Type)
 	return results, nil
 }
 
-func (app *fakeApplicator) WatchMatches(selector labels.Selector, labelType Type, quitCh chan struct{}) chan []Labeled {
+func (app *fakeApplicator) WatchMatches(selector labels.Selector, labelType Type, quitCh <-chan struct{}) chan []Labeled {
 	ch := make(chan []Labeled)
+	timer := time.NewTimer(time.Duration(0))
 	go func() {
 		for {
+			select {
+			case <-quitCh:
+				return
+			case <-timer.C:
+			}
+			// upper bound on request rate
+			timer.Reset(100 * time.Millisecond)
+
 			res, _ := app.GetMatches(selector, labelType)
 			select {
 			case <-quitCh:
 				return
 			case ch <- res:
 			}
+
 		}
 	}()
 	return ch
 }
 
+func (app *fakeApplicator) WatchMatchDiff(
+	selector labels.Selector,
+	labelType Type,
+	quitCh <-chan struct{},
+) <-chan *LabeledChanges {
+	outCh := make(chan *LabeledChanges)
+
+	go func() {
+		defer close(outCh)
+
+		var labelsList []Labeled
+		inCh := app.WatchMatches(selector, labelType, quitCh)
+
+		// Get a starting value to compare to
+		for {
+			select {
+			case <-quitCh:
+				return
+			case labelsList = <-inCh:
+				if labelsList == nil {
+					fmt.Println("Unexpected nil value received from WatchMatches, recovered")
+					continue
+				}
+			}
+			break
+		}
+
+		oldLabels := make(map[string]Labeled)
+		for _, labeled := range labelsList {
+			oldLabels[labeled.ID] = labeled
+		}
+
+		for {
+			var results []Labeled
+			select {
+			case <-quitCh:
+				return
+			case results = <-inCh:
+				if results == nil {
+					fmt.Println("Unexpected nil value received from WatchMatches, recovered")
+					continue
+				}
+			}
+
+			newLabels := make(map[string]Labeled)
+			for _, labeled := range results {
+				newLabels[labeled.ID] = labeled
+			}
+
+			outgoingChanges := &LabeledChanges{}
+			for id, labeled := range newLabels {
+				if _, ok := oldLabels[id]; !ok {
+					// If it is not observed, then it was created
+					outgoingChanges.Created = append(outgoingChanges.Created, Labeled{
+						ID:        id,
+						LabelType: labelType,
+						Labels:    copySet(labeled.Labels),
+					})
+					oldLabels[id] = labeled
+
+				} else if oldLabels[id].Labels.String() != labeled.Labels.String() {
```

Comment:
nit: it seems like there's a lot of code-reuse here. Given that it's in a fake implementation I don't care that much but consider extracting this to a function that both real and fake can use